installdir/hlpurge.1.in
.TH hldig 1 "January 2004"
.\" NAME should be all caps, SECTION should be 1-8, maybe w/ subsection
.\" other parms are allowed: see man(7), man(1)
.SH NAME
hlpurge \-  remove unused odocuments from the database (general maintenance script)
.SH SYNOPSIS
.B hlpurge
.I "[-][-a][-c configfile][-u][-v]"
.SH "DESCRIPTION"
hlpurge functions to remove specified URLs from the databases as well as bad URLs, 
unretrieved URLs, obsolete documents, etc. It is recommended that hlpurge be run 
after hldig to clean out any documents of this sort. 
.TP
.SH OPTIONS
.TP
.B \-
Take \fIURL\fR list from standard input (rather than specified with \fI-u\fR). Format of input file is one URL per line.
.B \-a
Use alternate work files. Tells hlpurge to append
.I .work
to database files, causing a second copy of the database to be
built. This allows the original files to be used by hlsearch
during the run. 
.TP
.B \-c \fIconfigfile\fR
Use the specified
.I configfile
instead of the default. 
.TP
.B \-u \fIURL\fR
Add this \fIURL\fR to the list of documents to remove. Must be specified multiple times if
more than one \fIURL\fR are to be removed. Should nor be used together with \fI-\fR.
.TP
.B \-v
Verbose mode. This increases the verbosity of the program.
Using more than 2 is probably only useful for debugging
purposes. The default verbose mode (using only one \fI\-v\fR) gives
a nice progress report while digging. 
.SH "FILES"
.TP
.B @DEFAULT_CONFIG_FILE@
The default configuration file.
.SH "SEE ALSO"
Please refer to the HTML pages (in the hldig-doc package)
.B /usr/share/doc/hldig-doc/html/index.html
and the manual pages
.I hldigconfig(8)
,
.I hldig(1)
and
.I hlmerge(1)
for a detailed description of hl://Dig and its commands.

.SH AUTHOR
This manual page was written by Robert Ribnitz, 
based on the HTML documentation of hl://Dig.


ChangeLog
2018-04-04  Johnny A. Solbu <johnny@solbu.net>

	* Remove broken build test link
	* rename ht -> hl in topdir files
	* Complete rename in installdir
	* Autoreconf
	* Apparently several subscripts still use the VERSION file. Should
	fix that sometime
	* Rename mans in configure and make scripts.
	* Rename man files and their contents

2018-04-02  Johnny A. Solbu <johnny@solbu.net>

	* VERSION file is obsolete now that version is in configure.ac
	[skip ci]
	* Update pot file: copyright are assigned to «the hl://Dig
	group» [skip ci]
	* commit fb7ea686ffe3fd4aac7e823f1855fb34a3f68dd3 Author: Johnny
	A. Solbu <johnny@solbu.net> Date:   Mon Apr 2 13:31:43 2018 +0200

2018-03-27  andy5995 <andy400-dev@yahoo.com>

	* dbbench.cc:fix conversion warning This should fix the warning when `make dist` is run: dbbench.cc:269:20: warning: ISO C++ forbids converting a string
	constant to ‘char*’ [-Wpedantic]    char
	*progname = "dbbench problem...";

2018-03-28  Johnny A. Solbu <johnny@solbu.net>

	* bump version: 0.2.0
	* Add bug address to configure
	* Put version in the configure.ac file, where it belongs
	* Update build tests location
	* Rename url in the last files

2018-03-27  Johnny A. Solbu <johnny@solbu.net>

	* Rename urls in more files
	* Change MSGID_BUGS_ADDRESS, point to the issue tracker
	* Update ChangeLog.md with the new URLs
	* New project management: Rename URLs
	* Don't track compiled translations in revision controll
	* Remove unneeded preceding slash

2018-03-27  andy5995 <andy400-dev@yahoo.com>

	* change htdig.gif to logo.jpg (#56) After changes by @jamessouth, I found a few other areas I forgot to
	mention to him.

2018-03-15  jamessouth <jamesdanielsouth@gmail.com>

	* issue 56 new logo and modified html files

2018-03-15  Andy Alt <andy5995@users.noreply.github.com>

	* Changlog.md:add note about default install paths [skip ci]
	* Makefile.am:remove CVS dist-hook [skip ci]
	* CONTRIBUTING.md:add note about "skip ci" [skip ci]
	* TESTING.md:updated [skip ci] This reflects changes to the new install locations, which was a
	recent patch submitted by @Solbu
	* push regenerated build scripts and makefiles Regenerated due to previous commit
	* configure.ac:change path of www images [skip ci] When installdir/Makefile is run, it uses the value here to prefix
	urls to images in the installdir/html templates with /www/ To be accessible, a symlink can be put in the web root that points
	to that directory.

2018-03-15  Johnny A. Solbu <github@solbu.net>

	* Install things in the proper places (#90)
	  Install things in PREFIX/share/hldir, rather than /var/www

2018-03-12  Andy Alt <andy5995@users.noreply.github.com>

	* Merge pull request #101 from solbu/installer-fixes Installer fixes (Issue #99)
	* republish site from changes in bcdf311 This republishes the site using changes @jamessouth made in
	bcdf3110c78a308404431f3f3a24d33503d283a5 (#81)

2018-03-12  james south <jamessouth@users.noreply.github.com>

	* make sure home page validates (#100) (fixes #81) [skip ci]

2018-03-12  Johnny A. Solbu <johnny@solbu.net>

	* Add ABOUT-NLS to distributed files, indent traling backslash
	* Add missing gettext.h header

2018-03-11  andy5995 <andy400-dev@yahoo.com>

	* ABOUT-NLS:add to distribution [skip ci]
	* remove '#' (forgot to do this in a previous commit) [skip ci]
	* .travis.yml:drop '#' from hldig Freenode channel [skip ci]
	* simplecgen:refactor trim() function
	* list #hldig for the chat room, not ##hldig [skip ci]
	* .travis.yml: change failure and error notifications Only notify on_failure if there's a change Always notify on error
	* .travis.yml:change irc notification options
	* .travis.yml:change osx and Xcode version (#98)
	* .travis.yml:change osx and Xcode version
	* .travis.yml:add matrix closes #76

2018-03-11  Johnny A. Solbu <github@solbu.net>

	* Fix test dir (issue #91) (#97) / Thanks @solbu
	* Fix «make dist» Some tools where renamed at some point but the test commands using
	them where not updated
	* Update test/test_function.in with the renamed filenames

2018-03-09  Andy Alt <andy5995@users.noreply.github.com>

	* Merge pull request #92 from solbu/doc-changes [skip ci] Simplify translation instructions

2018-03-07  andy5995 <andy400-dev@yahoo.com>

	* translation:add ru to LINGUAS, generate gmo file (closes #68)
	* travis.yml:add osx (#76)
	* ne.po:remove empty file [skip ci]

2018-03-07  Svitlana Galianova <svitlana.galianova@gmail.com>

	* added russian translation

2018-03-07  Johnny A. Solbu <github@solbu.net>

	* Uninstall fix (#84) (closes #48)
	* Downgrade required gettext version, to match my system
	* Add Solbu to authors list

2018-03-06  andy5995 <andy400-dev@yahoo.com>

	* ChangeLog.md:note about html docs not being installed anymore

2018-03-06  Johnny A. Solbu <johnny@solbu.net>

	* Run autoreconf
	* Don't install documentaion in docdir RPM packagers prefer to control that dir using a «%doc» macro
	* Ignore .patch files To avoid committing patches by mistake, when one is also working on
	packaging

2018-01-21  andy5995 <andy400-dev@yahoo.com>

	* fix formatting [skip ci]
	* remove Slack, use IRC found some leftovers to remove (see previous commit)
	* remove Slack workspace from contact info [skip ci] IRC should be sufficient for this project.  http://webchat.freenode.net?channels=%23%23hldig&uio=d4

2018-01-15  andy5995 <andy400-dev@yahoo.com>

	* docs/README: explain how to view the html files

2018-01-11  Andy Alt <andy5995@users.noreply.github.com>

	* [#24]installdirs:update search result html, add css (#82)
	* [#24]installdirs:update search result html, add css This will copy the same style.css that the web site uses. That seems
	easier than maintaining 2 separate css files.  I added a script that runs tidy-html5, but the user must install the
	program <http://www.htacg.org/> to use it. 
	* hlsearch results html pages updated [#24] See notes in the

hltools/hlpurge.cc
//
// htpurge.cc
//
// htpurge: A utility to remove specified URLs and any documents
// marked for removal from the word and document databases.
//
// Part of the ht://Dig package   <http://www.htdig.org/>
// Copyright (c) 1999-2004 The ht://Dig Group
// For copyright details, see the file COPYING in your distribution
// or the GNU Library General Public License (LGPL) version 2 or later
// <http://www.gnu.org/copyleft/lgpl.html>
//
// $Id: htpurge.cc,v 1.6 2004/05/28 13:15:25 lha Exp $
//

#ifdef HAVE_CONFIG_H
#include "htconfig.h"
#endif /* HAVE_CONFIG_H */

#include <fcntl.h>

#include "WordContext.h"
#include "HtWordReference.h"
#include "HtConfiguration.h"
#include "DocumentDB.h"
#include "DocumentRef.h"
#include "defaults.h"
#include "HtURLCodec.h"
#include "messages.h"

#include <errno.h>

#ifndef _MSC_VER                /* _WIN32 */
#include <unistd.h>
#endif

// If we have this, we probably want it.
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#elif HAVE_GETOPT_LOCAL
#include <getopt_local.h>
#endif

int verbose = 0;

Dictionary *purgeDocs (Dictionary *);
void purgeWords (Dictionary *);
void usage ();
void reportError (const char *msg);

//*****************************************************************************
// int main(int ac, char **av)
//
int
main (int ac, char **av)
{
  int alt_work_area = 0;
  String configfile = DEFAULT_CONFIG_FILE;
  int c;
  extern char *optarg;
  Dictionary *discard_ids = 0;
  Dictionary *discard_urls = new Dictionary;

  while ((c = getopt (ac, av, "vc:au:")) != -1)
  {
    switch (c)
    {
    case 'c':
      configfile = optarg;
      break;
    case 'v':
      verbose++;
      break;
    case 'a':
      alt_work_area++;
      break;
    case 'u':
      discard_urls->Add (optarg, NULL);
      break;
    case '?':
      usage ();
      break;
    }
  }

  HtConfiguration *config = HtConfiguration::config ();
  config->Defaults (&defaults[0]);

  if (access ((char *) configfile, R_OK) < 0)
  {
    reportError (form (_("Unable to find configuration file '%s'"),
                       configfile.get ()));
  }

  config->Read (configfile);

  //
  // Check url_part_aliases and common_url_parts for
  // errors.
  String url_part_errors = HtURLCodec::instance ()->ErrMsg ();

  if (url_part_errors.length () != 0)
    reportError (form (_("Invalid url_part_aliases or common_url_parts: %s"),
                       url_part_errors.get ()));

  if (alt_work_area != 0)
  {
    String configValue;

    configValue = config->Find ("word_db");
    if (configValue.length () != 0)
    {
      configValue << ".work";
      config->Add ("word_db", configValue);
    }

    configValue = config->Find ("doc_db");
    if (configValue.length () != 0)
    {
      configValue << ".work";
      config->Add ("doc_db", configValue);
    }

    configValue = config->Find ("doc_index");
    if (configValue.length () != 0)
    {
      configValue << ".work";
      config->Add ("doc_index", configValue);
    }

    configValue = config->Find ("doc_excerpt");
    if (configValue.length () != 0)
    {
      configValue << ".work";
      config->Add ("doc_excerpt", configValue);
    }
  }

  if (optind < ac && strcmp (av[optind], "-") == 0)
  {
    String str;
    while (!cin.eof ())
    {
      cin >> str;
      str.chop ("\r\n");
      if (str.length () > 0)
        discard_urls->Add (str, NULL);
    }
  }

  WordContext::Initialize (*config);

  // We pass in our list of URLs (which may be empty)
  // and we get back the list of IDs purged from the doc DB
  // to make sure words with these IDs are purged
  discard_ids = purgeDocs (discard_urls);
  delete discard_urls;
  discard_urls = 0;

  purgeWords (discard_ids);
  delete discard_ids;
  discard_ids = 0;

  return 0;
}

//*****************************************************************************
// Dictionary purgeDocs(Dictionary &purgeURLs)
// Pass in a hash of the URLs to delete (it could be empty)
// Return a hash of the IDs deleted from the doc DB
//
Dictionary *
purgeDocs (Dictionary * purgeURLs)
{
  HtConfiguration *config = HtConfiguration::config ();
  const String doc_db = config->Find ("doc_db");
  const String doc_index = config->Find ("doc_index");
  const String doc_excerpt = config->Find ("doc_excerpt");
  int remove_unused;
  int remove_unretrieved;
  DocumentDB db;
  List *IDs;
  int document_count = 0;
  Dictionary *discard_list = new Dictionary;

  //
  // Start the conversion by going through all the URLs that are in
  // the document database
  //
  if (db.Open (doc_db, doc_index, doc_excerpt) != OK)
    return discard_list;        // It's empty right now

  IDs = db.DocIDs ();

  if (IDs->Count () == 0)
    reportError (_("Database is empty!"));

  IDs->Start_Get ();
  IntObject *id;
  String idStr;
  String url;
  URL u_url;

  while ((id = (IntObject *) IDs->Get_Next ()))
  {
    DocumentRef *ref = db[id->Value ()];

    if (!ref)
      continue;

    db.ReadExcerpt (*ref);
    url = ref->DocURL ();
    u_url = URL ((char *) url);

    remove_unused =
      config->Boolean ("server", u_url.host (), "remove_bad_urls");
    remove_unretrieved =
      config->Boolean ("server", u_url.host (), "remove_unretrieved_urls");
    idStr = 0;
    idStr << id->Value ();

    if (ref->DocState () == Reference_noindex)
    {
      // This document either wasn't found or shouldn't be indexed.
      db.Delete (ref->DocID ());
      if (verbose)
        cout << _("Deleted, noindex: ID: ") << idStr << " URL: " << url << endl;
      discard_list->Add (idStr.get (), NULL);
    }
    else if (ref->DocState () == Reference_obsolete)
    {
      // This document was replaced by a newer one
      db.Delete (ref->DocID ());
      if (verbose)
        cout << _("Deleted, obsolete: ID: ") << idStr << " URL: " << url << endl;
      discard_list->Add (idStr.get (), NULL);
    }
    else if (remove_unused && ref->DocState () == Reference_not_found)
    {
      // This document wasn't actually found
      db.Delete (ref->DocID ());
      if (verbose)
        cout << "Deleted, not found: ID: " << idStr << " URL: "
          << url << endl;
      discard_list->Add (idStr.get (), NULL);
    }
    else if (remove_unused && strlen (ref->DocHead ()) == 0
             && ref->DocAccessed () != 0)
    {
      // For some reason, this document was retrieved, but doesn't
      // have an excerpt (probably because of a noindex directive)
      db.Delete (ref->DocID ());
      if (verbose)
        cout << _("Deleted, no excerpt: ID: ") << idStr << " URL:  "
          << url << endl;
      discard_list->Add (idStr.get (), NULL);
    }
    else if (remove_unretrieved && ref->DocAccessed () == 0)
    {
      // This document has not been retrieved
      db.Delete (ref->DocID ());
      if (verbose)
        cout << _("Deleted, never retrieved: ID: ") << idStr << " URL:  "
          << url << endl;
      discard_list->Add (idStr.get (), NULL);
    }
    else if (purgeURLs->Exists (url))
    {
      // This document has been marked to be purged by the user
      db.Delete (ref->DocID ());
      if (verbose)
        cout << _("Deleted, marked by user input: ID: ") << idStr << " URL: "
          << url << endl;
      discard_list->Add (idStr.get (), NULL);
    }
    else
    {
      // This is a valid document. Let's keep stats on it.
      if (verbose > 1)
        cout << "ID: " << idStr << " URL: " << url << endl;

      document_count++;
      if (verbose && document_count % 10 == 0)
      {
        cout << "hlpurge: " << document_count << '\n';
        cout.flush ();
      }
    }
    delete ref;
  }
  if (verbose)
    cout << "\n";

  delete IDs;
  db.Close ();

  return discard_list;
}

//
// Callback data dedicated to Dump and dump_word communication
//
class DeleteWordData:public Object
{
public:
  DeleteWordData (const Dictionary & discard_arg):discard (discard_arg)
  {
    deleted = remains = 0;
  }

  const Dictionary & discard;
  int deleted;
  int remains;
};

//*****************************************************************************
//
//
static int
delete_word (WordList * words, WordDBCursor & cursor,
             const WordReference * word_arg, Object & data)
{
  const HtWordReference *word = (const HtWordReference *) word_arg;
  DeleteWordData & d = (DeleteWordData &) data;
  static String docIDStr;

  docIDStr = 0;
  docIDStr << word->DocID ();

  if (d.discard.Exists (docIDStr))
  {
    if (words->Delete (cursor) != 0)
    {
      cerr << "hlpurge: deletion of " << (char *) word->
        Get () << " failed " << strerror (errno) << "\n";
      return NOTOK;
    }
    if (verbose)
    {
      cout << "hlpurge: Discarding ";
      if (verbose > 2)
        cout << (char *) word->Get ();
      else
        cout << word->Word ();
      cout << "\n";
      cout.flush ();
    }
    d.deleted++;
  }
  else
  {
    d.remains++;
  }

  return OK;
}

//*****************************************************************************
// void purgeWords(Dictionary *discard_list)
//
void
purgeWords (Dictionary * discard_list)
{
  HtConfiguration *config = HtConfiguration::config ();
  HtWordList words (*config);
  DeleteWordData data (*discard_list);

  words.Open (config->Find ("word_db"), O_RDWR);
  WordCursor *search = words.Cursor (delete_word, &data);
  search->Walk ();
  delete search;

  words.Close ();

  if (verbose)
    cout << "\n";

}

//*****************************************************************************
// void usage()
//   Display program usage information
//
void
usage ()
{
  Usage help;
  cout << _("usage:");
  cout << " hlpurge [-][-u url][-v][-a][-c configfile]\n";
  printf (_("This program is part of hl://Dig %s\n\n"), VERSION);
  cout << _("Options:\n");
  cout <<
    "\t-\tURL input. Read in a list of URLs to remove, one per line.\n\n";
  cout << "\t-u\tURL input. Add this url to the list of URLs to remove.\n";
  cout << "\t\t(can be specified multiple times)\n\n";

  help.verbose ();

  help.alternate_common ();
  cout << "\t\tTells hlpurge to append .work to the database files \n";
  cout << "\t\tallowing it to operate on a second set of databases.\n\n";
  help.config ();

  exit (0);
}


//*****************************************************************************
// Report an error and die
//
void
reportError (const char *msg)
{
  cout << "hlpurge: " << msg << "\n\n";
  exit (1);
}
